packages/auth/AuthStore.ts

```diff
@@ -0,0 +1,126 @@
+
+interface AuthStoreProps {
+  authSettings: Auth0ClientOptions | undefined;
+}
+export class AuthStore {
+  readonly authSettings: Auth0ClientOptions | undefined;
+
+  private authClient: Auth0Client | undefined;
+
+  emailVerified?: boolean;
+
+  isAuthorized: boolean;
+
+  isLoading: boolean;
+
+  user?: User;
+
+  constructor({ authSettings }: AuthStoreProps) {
+    makeAutoObservable(this);
+
+    this.authSettings = authSettings;
+    this.authClient = undefined;
+    this.isAuthorized = false;
+    this.isLoading = true;
+  }
+
+  private get auth0Client(): Promise<Auth0Client> {
+    return createAuth0Client(this.authSettings as Auth0ClientOptions);
+  }
+
+  async authenticate(): Promise<void> {
+    const auth0 = await this.auth0Client;
+
+    runInAction(() => {
+      this.authClient = auth0;
+    });
+
+    /**
+     * @remarks
+     *
+     * Auth0 library responds to the presence of query parameters (code, state, and error) all of which
+     * indicate that we have been redirected from the Auth0 login page. If they are not removed before
+     * further interaction with the library, the application enters an infinite loop with newly regenerated
+     * state/code/error params. The below logic mitigates that by stripping away the parameters
+     * completely or replaces them with the `targetUrl`.
+     */
+    const urlQuery = qs.parse(window.location.search, {
+      ignoreQueryPrefix: true,
+    });
+
+    if (urlQuery.code && urlQuery.state) {
+      const { appState } = await auth0.handleRedirectCallback();
+      // auth0 params are single-use, must be removed from history or they can cause errors
+      let replacementUrl;
+
+      if (appState && appState.targetUrl) {
+        replacementUrl = appState.targetUrl;
+      } else {
+        // strip away all query params just to be safe
+        replacementUrl = `${window.location.origin}${window.location.pathname}`;
+      }
+
+      window.history.replaceState({}, document.title, replacementUrl);
+    }
+
+    if (await auth0.isAuthenticated()) {
+      const user = await auth0.getUser();
+
+      runInAction(() => {
+        this.isLoading = false;
+        this.user = user;
+        this.isAuthorized = true;
+        this.emailVerified = Boolean(user?.email_verified);
+      });
+    } else {
+      await auth0.loginWithRedirect({
+        appState: { targetUrl: window.location.href },
+      });
+    }
+  }
+
+  async logoutUser() {
+    runInAction(() => {
+      this.isAuthorized = false;
+      this.isLoading = true;
+    });
+
+    return this.authClient?.logout({ returnTo: window.location.origin });
+  }
+
+  async getToken(options?: GetTokenSilentlyOptions): Promise<string | Error> {
+    try {
+      if (this.authClient) {
+        return this.authClient.getTokenSilently(options);
+      }
+      return Promise.reject(new Error("No auth client initialized."));
+    } catch (error) {
+      throw new Error(`Unable to retrieve token - ${error}`);
```

Comment:
maybe also return a rejected promise here for consistency?

Reply:
Sorry Ian, can you clarify this note?

Reply:
i just meant that one branch was rejecting a promise with an error while the other is throwing it, which seems weird? I would do one or the other in both cases (i guess it doesn't really matter which since the end result is the same in an async function. 

Unless there is some good reason why the behavior needs to differ, in which case I think a comment explaining it would suffice

Reply:
Thanks for clarifying offline!